### PR TITLE
[webgpu] Prepare for binary op restructuring

### PR DESCRIPTION
**Pull request by hujiajie**
The current code isn't great in that the vec4 shaders have diverged from the scalar ones more than necessary. The situation will be much improved by moving the NaN checking snippet in each op to a centralized place. Here is the common preparation work, so that following refactoring can be done on a per-op basis.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7257)

### CODE REVIEW

File: tfjs-backend-webgpu/src/binary_op_util.ts [hunk 1]
```diff
@@ -264,6 +318,10 @@ export function getBinaryOpString(
     case BinaryOpType.SUB:
       return SUB;
     default:
-      throw new Error(`BinaryType ${type} is not implemented!`);
+      // throw new Error(`BinaryType ${type} is not implemented!`);
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/binary_op_util.ts:321]: You may put NaN-insensitive ops in another do while, and keep the throw here.

File: tfjs-backend-webgpu/src/binary_op_util.ts [hunk 2]
```diff
@@ -216,6 +217,59 @@ function getBinaryWithNanString(op: string, useVec4: boolean) {
...
+        doOpSnippet = POW_VEC4;
+        break;
+      default:
+        continue;
+    }
+    const checkNaNSnippet =
+        useVec4 ? CHECK_NAN_SNIPPET_VEC4 : CHECK_NAN_SNIPPET;
+    return `
+      ${doOpSnippet}
+      ${checkNaNSnippet}
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/binary_op_util.ts:268]: This introduces a duplicated check that is not necessary.

**Review by gyagp — APPROVED** [review]
Personally I don't think this PR is necessary, and we don't need to code with follow-up refactoring per op. 
Just approve here to unblock the overall progress.

File: tfjs-backend-webgpu/src/binary_op_util.ts (hunk 2 above)
Comment (hujiajie) [reply tfjs-backend-webgpu/src/binary_op_util.ts:268]: _"That's right. It's part of the plan." -- The Dark Forest_

File: tfjs-backend-webgpu/src/binary_op_util.ts [hunk 3]
```diff
@@ -42,8 +42,9 @@ export enum BinaryOpType {
 }
 
 const CHECK_NAN_SNIPPET = `
-  if (isnan(a)) { return a; }
-  if (isnan(b)) { return b; }
+  if isNaN { return valueForNaN; }
+  if isnan(a) { return valueForNaN; }
+  if isnan(b) { return valueForNaN; }
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/binary_op_util.ts:47]: LGTM to only change above three lines to align with `CHECK_NAN_SNIPPET_VEC4`.
For other part, It seems that you are introducing more duplicated things. Like twice `checkNaNSnippet` check and some never running code. I understand you may clean the NaN ops to remove the extra `checkNaNSnippet` in following PRs. But I prefer you do it when everything is ready or do ops refactor first. I hope each PR is clean and don't bring some confusing things.

**Review by qjia7 — CHANGES_REQUESTED** [review]

File: tfjs-backend-webgpu/src/binary_op_util.ts (hunk 1 above)
Comment (hujiajie) [reply tfjs-backend-webgpu/src/binary_op_util.ts:321]: Will do in the final cleanup, but not possible now.

```javascript
do {
  switch (type) {
    case BinaryOpType.FOOBAR:
      return FOOBAR;
    default:
      continue;
  }
  return `blah blah`;  // <-- still unreachable
} while (false);
```

File: tfjs-backend-webgpu/src/binary_op_util.ts (hunk 3 above)
Comment (hujiajie) [reply tfjs-backend-webgpu/src/binary_op_util.ts:47]: > LGTM to only change above three lines to align with CHECK_NAN_SNIPPET_VEC4.

Will do in the final cleanup, but not possible now. Current scalar implementations are lacking the definition of `resultTemp`.

**Pull request closed without merging**